File: Python/Flyweight/PieceBoard.py
# -*- coding:utf-8 -*-

from Piece import *

class PieceBoard:

    def __init__(self, blackName, whiteName):
        self.__blackName = blackName
        self.__whiteName = whiteName
        self.__blackPiece = None
        self.__whitePiece = None
        self.__vecPiecePos = []

    def SetPiece(self, color, pos):
        if (color == PieceColor.BLACK):
            if(self.__blackPiece == None):
                self.__blackPiece = BlackPiece()
            print "%s%s%d %d"%(self.__blackName, "在位置", pos.x, pos.y)
            self.__blackPiece.Draw()
        else:
            if(self.__whitePiece == None):
                self.__whitePiece = WhitePiece()
            print "%s%s%d %d"%(self.__whiteName, "在位置", pos.x, pos.y)
            self.__whitePiece.Draw()
        self.__vecPiecePos.append(pos)

File: CPlusPlus/Flyweight/src/PieceBoard.cpp
/*************************************************************************
    > File Name: PieceBoard.cpp
    > Author: ma6174
    > Mail: ma6174@163.com 
    > Created Time: 2017年03月13日 星期一 13时28分08秒
 ************************************************************************/

#include"PieceBoard.h"
#include"WhitePiece.h"
#include"BlackPiece.h"
#include<iostream>

using namespace std;

PieceBoard::PieceBoard(const string& bName, const string& wName):blackName(bName), whiteName(wName)
{
	;
}

PieceBoard::~PieceBoard()
{
	;
}

void PieceBoard::SetPiece(const PieceColor& color, const PiecePos& piecePos)
{
	if(color == BLACK)
	{
		if(!blackPiece)
		{
			blackPiece = make_shared<BlackPiece>();
		}
		cout<<blackName<<"在位置"<<piecePos.x<<" "<<piecePos.y<<endl;
		blackPiece->Draw();
	}
	else
	{
		if(!whitePiece)
		{
			whitePiece = make_shared<WhitePiece>();
		}
		cout<<whiteName<<"在位置"<<piecePos.x<<" "<<piecePos.y<<endl;
		whitePiece->Draw();

	}
	vecPiecePos.push_back(piecePos);
}
